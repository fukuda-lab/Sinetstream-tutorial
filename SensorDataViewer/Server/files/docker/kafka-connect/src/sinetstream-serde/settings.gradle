rootProject.name = 'sinetstream-kafka-serde'
