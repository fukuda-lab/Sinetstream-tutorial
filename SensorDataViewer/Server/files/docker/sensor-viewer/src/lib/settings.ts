import { format, parseISO, sub } from 'date-fns';
import { persisted } from 'svelte-local-storage-store';
import {
  derived, writable, type Readable, type Writable,
} from 'svelte/store';
import * as zod from 'zod';

const paneVisibleSchema = zod.object({
  map: zod.boolean().optional(),
  image: zod.boolean().optional(),
  perftool: zod.boolean().optional(),
  lte: zod.boolean().optional(),
  sensors: zod.array(zod.string()).optional(),
});
const playerSchema = zod.object({
  from: zod.union([zod.preprocess(
    (arg) => (typeof arg === 'string' ? parseISO(arg) : arg),
    zod.date(),
  ), zod.string().regex(/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}$/)]),
  to: zod.union([zod.preprocess(
    (arg) => (typeof arg === 'string' ? parseISO(arg) : arg),
    zod.date(),
  ), zod.string().regex(/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}$/)]),
  tick: zod.number().gte(100),
  speed: zod.number().gte(1).lte(100),
  repeat: zod.boolean(),
});
const chartV0Schema = zod.object({
  minSize: zod.object({
    width: zod.number().gte(200).optional(),
    height: zod.number().gte(150).optional(),
  }).optional(),
  line: zod.object({
    enabled: zod.boolean().optional(),
    size: zod.number().gt(0).optional(),
  }).optional(),
  point: zod.object({
    enabled: zod.boolean().optional(),
    size: zod.number().gt(0).optional(),
  }).optional(),
});
const layoutTemplateSchema = zod.enum(['horizontal', 'vertical-horizontal', 'vertical', 'vertical-2']);
const viewerSettingSchemaV0 = zod.object({
  name: zod.string(),
  publisher: zod.string().optional(),
  imageTopic: zod.string().optional(),
  perftoolName: zod.string().optional(),
  panes: paneVisibleSchema.optional(),
  imageUpdate: zod.enum(['push', 'polling']).optional(),
  player: playerSchema.optional(),
  layout: zod.object({ template: layoutTemplateSchema }).optional(),
  chart: chartV0Schema.optional(),
});
const androidSchema = zod.object({
  publisher: zod.string(),
  map: zod.boolean(),
  lte: zod.boolean(),
  sensors: zod.array(zod.string()),
});
const picameraSchema = zod.object({
  topic: zod.string(),
  update: zod.enum(['push', 'polling']),
  visible: zod.boolean(),
});
const layoutSchema = zod.object({
  template: layoutTemplateSchema,
});
const perftoolSchema = zod.object({
  name: zod.string(),
  visible: zod.boolean(),
});
const chartV1Schema = zod.object({
  minSize: zod.object({
    width: zod.number().gte(200),
    height: zod.number().gte(150),
  }),
  line: zod.object({
    enabled: zod.boolean(),
    size: zod.number().gt(0),
  }),
  point: zod.object({
    enabled: zod.boolean(),
    size: zod.number().gt(0),
  }),
});
const viewerSettingSchemaV1 = zod.object({
  name: zod.string(),
  version: zod.literal('1.0'),
  android: androidSchema,
  chart: chartV1Schema,
  image: picameraSchema,
  layout: layoutSchema,
  perftool: perftoolSchema,
  player: playerSchema,
});

type ViewerSettingV0 = zod.infer<typeof viewerSettingSchemaV0>;
type ViewerSettingV1 = zod.infer<typeof viewerSettingSchemaV1>;
type ViewerSetting = ViewerSettingV0 | ViewerSettingV1;
type PaneVisible = zod.infer<typeof paneVisibleSchema>;
type SensorViewerMode = 'viewer' | 'player';
type LayoutType = 'horizontal' | 'vertical-horizontal' | 'vertical' | 'vertical-2';
type ImageUpdateMethod = 'push' | 'polling';

function isViewerSettingV0(value: unknown): value is ViewerSettingV0 {
  return viewerSettingSchemaV0.safeParse(value).success;
}

function isViewerSettingV1(value: unknown): value is ViewerSettingV1 {
  return viewerSettingSchemaV1.safeParse(value).success;
}

const mode: Writable<SensorViewerMode> = persisted('mode', 'viewer');
const settings: Writable<ViewerSetting[]> = persisted('settings', []);
const settingsIndex = writable(0);

function defaultDateTime(minutes = 0): string {
  return format(sub(new Date(), { minutes }), 'yyyy-MM-dd HH:mm');
}

function toV1(v0: ViewerSettingV0): ViewerSettingV1 {
  return {
    name: v0.name,
    version: '1.0',
    android: {
      publisher: v0.publisher ?? '',
      map: v0.panes?.map ?? true,
      lte: v0.panes?.lte ?? true,
      sensors: v0.panes?.sensors ?? [],
    },
    chart: {
      minSize: {
        width: v0.chart?.minSize?.width ?? 310,
        height: v0.chart?.minSize?.height ?? 250,
      },
      line: {
        enabled: v0.chart?.line?.enabled ?? true,
        size: v0.chart?.line?.size ?? 1,
      },
      point: {
        enabled: v0.chart?.point?.enabled ?? true,
        size: v0.chart?.point?.size ?? 2,
      },
    },
    image: {
      topic: v0.imageTopic ?? '',
      update: v0.imageUpdate ?? 'push',
      visible: v0.panes?.image ?? true,
    },
    layout: {
      template: v0.layout?.template ?? 'horizontal',
    },
    perftool: {
      name: v0.perftoolName ?? '',
      visible: v0.panes?.perftool ?? true,
    },
    player: {
      from: v0.player?.from ?? defaultDateTime(15),
      to: v0.player?.to ?? defaultDateTime(),
      tick: v0.player?.tick ?? 1200,
      speed: v0.player?.tick ?? 10,
      repeat: v0.player?.repeat ?? false,
    },
  };
}

function defaultSetting(settingName: string): ViewerSettingV1 {
  return toV1({ name: settingName });
}

const currentSetting: Readable<ViewerSettingV1> = derived(
  [settings, settingsIndex],
  ([$settings, $settingsIndex], set) => {
    if ($settingsIndex < $settings.length && $settingsIndex >= 0) {
      const current = $settings[$settingsIndex];
      if (isViewerSettingV1(current)) {
        set(current);
      } else if (isViewerSettingV0(current)) {
        set(toV1(current));
      }
    } else {
      set(defaultSetting(''));
    }
  },
);

const name: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.name);
});

const publisher: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.android.publisher);
});

const perftoolName: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.perftool.name);
});

const imageTopic: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.image.topic);
});

const pane: Readable<PaneVisible> = derived([currentSetting], ([$currentSetting], set) => {
  const current = $currentSetting;
  set({
    lte: current.android.lte,
    map: current.android.map,
    sensors: current.android.sensors,
    image: current.image.visible,
    perftool: current.perftool.visible,
  });
});

const charts: Readable<string[]> = derived([currentSetting], ([$currentSetting], set) => {
  const current = $currentSetting;
  const { sensors } = current.android;
  const result: string[] = [...sensors];
  if (current.android.lte) {
    result.unshift('lte');
  }
  if (current.perftool.visible) {
    result.unshift('perftool');
  }
  set(result);
});

const imageUpdateMethod: Readable<ImageUpdateMethod> = derived(
  [currentSetting],
  ([$currentSetting], set) => {
    set($currentSetting.image.update);
  },
);

export {
  settings,
  settingsIndex,
  mode,
  name,
  publisher,
  perftoolName,
  imageTopic,
  pane,
  charts,
  imageUpdateMethod,
  currentSetting,
  isViewerSettingV0,
  isViewerSettingV1,
  viewerSettingSchemaV1,
};
export type {
  SensorViewerMode,
  ImageUpdateMethod,
  ViewerSetting,
  ViewerSettingV0,
  ViewerSettingV1,
  PaneVisible,
  LayoutType,
};
