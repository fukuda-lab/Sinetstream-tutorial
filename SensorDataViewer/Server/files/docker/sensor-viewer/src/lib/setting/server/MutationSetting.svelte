<script lang="ts">
  import {
    getContextClient,
    gql,
    mutationStore,
    queryStore,
  } from "@urql/svelte";
  import { createEventDispatcher } from "svelte";
  import type { ViewerSettingV1 } from "../../settings";
  import { currentSetting } from "../../settings";
  import { pickupParameter } from "./server";

  export let name: string | null;
  export let comment: string | null = "";
  export let targets: string[] = [];

  const client = getContextClient();
  $: queryConfig = queryStore({
    client,
    query: gql`
      query ($name: String!) {
        viewer_config(where: { name: { _eq: $name } }) {
          id
          comment
          config
        }
      }
    `,
    variables: { name: name ?? "" },
  });

  let mutationResult = undefined;
  const insertConfig = () => {
    if ($mutationResult != null) {
      return;
    }
    const config = pickupParameter($currentSetting, targets);
    mutationResult = mutationStore({
      client,
      query: gql`
        mutation ($config: jsonb, $name: String, $comment: String) {
          insert_viewer_config(
            objects: { config: $config, name: $name, comment: $comment }
          ) {
            returning {
              id
              name
              config
              comment
            }
          }
        }
      `,
      variables: { name, config, comment },
    });
  };
  const updateConfig = (id: number, cfg: ViewerSettingV1, cmt: string) => {
    if ($mutationResult != null) {
      return;
    }
    const newComment = comment?.length > 0 ? comment : cmt;
    const config = pickupParameter($currentSetting, targets, cfg);
    mutationResult = mutationStore({
      client,
      query: gql`
        mutation ($id: Int!, $config: jsonb, $comment: String) {
          update_viewer_config_by_pk(
            pk_columns: { id: $id }
            _set: { config: $config, comment: $comment }
          ) {
            id
            name
            config
            comment
          }
        }
      `,
      variables: { id, config, comment: newComment },
    });
  };

  let dialog = false;
  let force = false;
  $: if (name != null && $queryConfig?.data) {
    const q = $queryConfig?.data?.viewer_config;
    if (q.length > 0) {
      if (force) {
        dialog = false;
        const { id, config, comment } = q[0];
        updateConfig(id, config, comment);
      } else {
        dialog = true;
      }
    } else {
      insertConfig();
    }
  }

  const dispatch = createEventDispatcher();
  $: if ($queryConfig?.error != null) {
    dispatch("notify", { error: $queryConfig.error });
  }
  $: if ($mutationResult?.error != null) {
    dispatch("notify", { error: $mutationResult.error });
  } else if ($mutationResult?.data != null) {
    dispatch("notify", { mutation: $mutationResult.data });
  }

  function onCancel() {
    dialog = false;
    dispatch("cancel");
  }

  function forceUpdate() {
    force = true;
  }
</script>

{#if dialog}
  <div class="overlay active" />
  <div class="modal active">
    <h6>Warning</h6>
    <div>
      <i>warning</i>
      <span>
        サーバに登録されている設定を上書き更新しようとしています。
        よろしいですか？
      </span>
    </div>
    <nav class="right-align">
      <button class="border" on:click={onCancel}>Cancel</button>
      <button on:click={forceUpdate}>OK</button>
    </nav>
  </div>
{/if}
