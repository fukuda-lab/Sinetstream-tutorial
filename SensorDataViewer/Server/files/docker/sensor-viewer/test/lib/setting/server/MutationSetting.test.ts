import { act, cleanup, render } from '@testing-library/svelte';
import userEvent from '@testing-library/user-event';
import * as urql from '@urql/svelte';
import { writable } from 'svelte/store';
import {
  afterEach, beforeEach, describe, expect, it, vi,
} from 'vitest';
import MutationSetting from '../../../../src/lib/setting/server/MutationSetting.svelte';
import { settings, settingsIndex } from '../../../../src/lib/settings';

describe('MutationSetting.svelt', () => {
  const resultQuery = writable({ fetching: false, data: undefined, error: undefined });
  const resultMutation = writable({ fetching: false, data: undefined, error: undefined });
  let user;

  beforeEach(() => {
    user = userEvent.setup();
    vi.mock('@urql/svelte');
    settings.set([]);
    settingsIndex.set(0);
    resultQuery.set({ fetching: true, data: undefined, error: undefined });
    resultMutation.set({ fetching: true, data: undefined, error: undefined });
    vi.spyOn(urql, 'queryStore').mockReturnValue(resultQuery);
    vi.spyOn(urql, 'mutationStore').mockReturnValue(resultMutation);
  });

  afterEach(() => cleanup());

  describe('mutation', () => {
    it('insert', async () => {
      const { component, container } = render(MutationSetting, {
        props: {
          name: 'setting-0',
          comment: 'comment 0',
          targets: ['android', 'image'],
        },
      });
      const handler = vi.fn();
      component.$on('notify', handler);

      await act(() => {
        resultQuery.update((x) => {
          const data = { viewer_config: [] };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      expect(handler).not.toHaveBeenCalled();
      expect(container.innerHTML).toMatchSnapshot();

      await act(() => {
        resultMutation.update((x) => {
          const data = { insert_viewer_config: { returning: [{ id: 17 }] } };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      expect(handler).toHaveBeenCalledOnce();
      const { detail } = handler.mock.calls[0][0];
      expect(detail).toHaveProperty('mutation');
      expect(detail).not.toHaveProperty('error');
    });

    it('update', async () => {
      const { component, container, getByRole } = render(MutationSetting, {
        props: {
          name: 'setting-1',
          comment: 'comment 1',
          targets: ['android', 'image'],
        },
      });
      const handler = vi.fn();
      component.$on('notify', handler);

      await act(() => {
        resultQuery.update((x) => {
          const data = { viewer_config: [{ id: 17, config: {} }] };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      expect(handler).not.toHaveBeenCalled();
      expect(container.innerHTML).toMatchSnapshot();

      await user.click(getByRole('button', { name: 'OK' }));
      expect(container.innerHTML).toMatchSnapshot();

      await act(() => {
        resultMutation.update((x) => {
          const data = { update_viewer_config_by_pk: { id: 17 } };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      expect(handler).toHaveBeenCalledOnce();
      const { detail } = handler.mock.calls[0][0];
      expect(detail).toHaveProperty('mutation');
      expect(detail).not.toHaveProperty('error');
    });

    it('cancel update', async () => {
      const { component, container, getByRole } = render(MutationSetting, {
        props: {
          name: 'setting-1',
          comment: 'comment 1',
          targets: ['android', 'image'],
        },
      });
      const handler = vi.fn();
      component.$on('notify', handler);
      const cancelHandler = vi.fn();
      component.$on('cancel', cancelHandler);

      await act(() => {
        resultQuery.update((x) => {
          const data = { viewer_config: [{ id: 17, config: {} }] };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      expect(cancelHandler).not.toHaveBeenCalledOnce();
      expect(handler).not.toHaveBeenCalled();
      expect(container.innerHTML).toMatchSnapshot();

      await user.click(getByRole('button', { name: 'Cancel' }));
      expect(container.innerHTML).toMatchSnapshot();
      expect(cancelHandler).toHaveBeenCalledOnce();
      expect(handler).not.toHaveBeenCalledOnce();
    });
  });

  describe('error', () => {
    const ERROR_MESSAGE = '[Network] Failed to fetch';

    it('query error', async () => {
      const { component } = render(MutationSetting, {
        props: { name: 'setting-0' },
      });
      const handler = vi.fn();
      component.$on('notify', handler);

      await act(() => {
        resultQuery.update((x) => {
          const error = { message: ERROR_MESSAGE };
          const fetching = false;
          return { ...x, fetching, error };
        });
      });
      expect(handler).toHaveBeenCalledOnce();
      expect(handler.mock.calls[0][0]?.detail).toMatchObject({ error: { message: ERROR_MESSAGE } });
    });

    it('insert error', async () => {
      const { component } = render(MutationSetting, {
        props: { name: 'setting-0' },
      });
      const handler = vi.fn();
      component.$on('notify', handler);

      await act(() => {
        resultQuery.update((x) => {
          const data = { viewer_config: [] };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      expect(handler).not.toHaveBeenCalled();
      await act(() => {
        resultMutation.update((x) => {
          const error = { message: ERROR_MESSAGE };
          const fetching = false;
          return { ...x, fetching, error };
        });
      });
      expect(handler).toHaveBeenCalledOnce();
      expect(handler.mock.calls[0][0]?.detail).toMatchObject({ error: { message: ERROR_MESSAGE } });
    });

    it('update error', async () => {
      const { component, container, getByRole } = render(MutationSetting, {
        props: { name: 'setting-1' },
      });
      const handler = vi.fn();
      component.$on('notify', handler);

      await act(() => {
        resultQuery.update((x) => {
          const data = { viewer_config: [{ id: 17, config: {} }] };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      expect(handler).not.toHaveBeenCalled();
      expect(container.innerHTML).toMatchSnapshot();

      await user.click(getByRole('button', { name: 'OK' }));
      expect(container.innerHTML).toMatchSnapshot();

      await act(() => {
        resultMutation.update((x) => {
          const error = { message: ERROR_MESSAGE };
          const fetching = false;
          return { ...x, fetching, error };
        });
      });
      expect(handler).toHaveBeenCalledOnce();
      expect(handler.mock.calls[0][0]?.detail).toMatchObject({ error: { message: ERROR_MESSAGE } });
    });
  });
});
