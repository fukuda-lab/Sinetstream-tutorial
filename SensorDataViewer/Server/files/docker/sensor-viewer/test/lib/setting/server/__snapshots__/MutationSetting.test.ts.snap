// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MutationSetting.svelt > error > update error 1`] = `
"<div><div class=\\"overlay active\\"></div> <div class=\\"modal active\\"><h6>Warning</h6> <div><i>warning</i> <span>サーバに登録されている設定を上書き更新しようとしています。
        よろしいですか？</span></div> <nav class=\\"right-align\\"><button class=\\"border\\">Cancel</button> <button>OK</button></nav></div><!--<MutationSetting>--></div>"
`;

exports[`MutationSetting.svelt > error > update error 2`] = `"<div><!--<MutationSetting>--></div>"`;

exports[`MutationSetting.svelt > mutation > cancel update 1`] = `
"<div><div class=\\"overlay active\\"></div> <div class=\\"modal active\\"><h6>Warning</h6> <div><i>warning</i> <span>サーバに登録されている設定を上書き更新しようとしています。
        よろしいですか？</span></div> <nav class=\\"right-align\\"><button class=\\"border\\">Cancel</button> <button>OK</button></nav></div><!--<MutationSetting>--></div>"
`;

exports[`MutationSetting.svelt > mutation > cancel update 2`] = `"<div><!--<MutationSetting>--></div>"`;

exports[`MutationSetting.svelt > mutation > insert 1`] = `"<div><!--<MutationSetting>--></div>"`;

exports[`MutationSetting.svelt > mutation > update 1`] = `
"<div><div class=\\"overlay active\\"></div> <div class=\\"modal active\\"><h6>Warning</h6> <div><i>warning</i> <span>サーバに登録されている設定を上書き更新しようとしています。
        よろしいですか？</span></div> <nav class=\\"right-align\\"><button class=\\"border\\">Cancel</button> <button>OK</button></nav></div><!--<MutationSetting>--></div>"
`;

exports[`MutationSetting.svelt > mutation > update 2`] = `"<div><!--<MutationSetting>--></div>"`;
